src/symbols/analyzer.rs

```diff
@@ -2978,4 +3044,196 @@ impl<'a> SymbolFiller<'a> {
+  ) {
+    for param in params {
+      let (pat, default_value) = match param {
+        Pat::Assign(pat) => (&*pat.left, Some(&pat.right)),
+        _ => (param, None),
+      };
+
+      let type_ann = match pat {
+        Pat::Ident(n) => n.type_ann.as_ref(),
+        Pat::Array(n) => n.type_ann.as_ref(),
+        Pat::Object(n) => n.type_ann.as_ref(),
+        Pat::Rest(n) => n.type_ann.as_ref(),
+        Pat::Invalid(_) | Pat::Expr(_) | Pat::Assign(_) => unreachable!(),
+      };
+
+      if let Some(type_ann) = type_ann {
+        self.fill_ts_type(symbol_id, &type_ann.type_ann);
+      } else if let Some(default_value) = default_value {
+        self.fill_expr(symbol_id, default_value);
+      }
+
+      for ident in find_pat_ids::<_, Ident>(pat) {
+        self.add_param(symbol_id, ident);
+      }
+    }
+  }
+
+  fn add_param(&self, parent_id: SymbolId, ident: Ident) {
+    let range = ident.range();
+    self.builder.ensure_symbol_for_swc_id(
+      ident.to_id(),
+      SymbolDecl::new(
+        SymbolDeclKind::Definition(SymbolNode(SymbolNodeInner::Param {
+          ident,
+          source: self.source.clone(),
+        })),
+        range,
+      ),
+      parent_id,
+    );
+  }
+
+  fn fill_ts_type(&self, symbol_id: SymbolId, type_ann: &TsType) {
+    struct TsTypeVisitor<I: Fn(&Ident)> {
+      idents: I,
+    }
+    impl<I: Fn(&Ident)> Visit for TsTypeVisitor<I> {
+      fn visit_ts_infer_type(&mut self, n: &TsInferType) {
+        (self.idents)(&n.type_param.name);
+      }
+
+      fn visit_ts_type_param(&mut self, n: &TsTypeParam) {
+        (self.idents)(&n.name);
+      }
+
+      fn visit_ts_fn_param(&mut self, n: &TsFnParam) {
+        match n {
+          TsFnParam::Ident(ident) => {
+            (self.idents)(&ident.id);
+          }
+          TsFnParam::Array(n) => {
+            for ident in find_pat_ids::<_, Ident>(n) {
+              (self.idents)(&ident);
+            }
+          }
+          TsFnParam::Rest(n) => {
+            for ident in find_pat_ids::<_, Ident>(n) {
+              (self.idents)(&ident);
+            }
+          }
+          TsFnParam::Object(n) => {
+            for ident in find_pat_ids::<_, Ident>(n) {
+              (self.idents)(&ident);
+            }
+          }
+        }
+      }
+    }
+
+    let mut visitor = TsTypeVisitor {
+      idents: |ident| {
+        self.add_param(symbol_id, ident.clone());
+      },
+    };
+    type_ann.visit_with(&mut visitor);
+  }
+
+  fn fill_fn_body(
+    &self,
+    symbol_id: SymbolId,
+    body: &BlockStmt,
+    fn_kind: FunctionKind,
+  ) {
+    let analysis = analyze_return_stmts_in_function_body(body);
+    if let (
+      FunctionKind::DeclarationLike | FunctionKind::ExpressionLike,
+      ReturnStatementAnalysis::Single(stmt),
+    ) = (fn_kind, analysis)
+    {
+      self.fill_expr(symbol_id, stmt.arg.as_ref().unwrap());
+    }
+  }
+
+  fn fill_expr(&self, symbol_id: SymbolId, expr: &Expr) {
```

Comment:
Maybe add a comment here explaining why we need to fill expressions. I got confused by this initially.